Scan/STLcd/lcd_scan.c

```diff
@@ -338,11 +345,204 @@ inline void LCD_setup()
 	PORTC_PCR3 = PORT_PCR_SRE | PORT_PCR_DSE | PORT_PCR_MUX(4);
 }
 
+static uint32_t STLcdSyncStartMillis;
+static uint32_t STLcdSyncStartTicks;
+
+// Display buffer
+uint8_t STLcdBuffer[LCD_PAGE_LEN * LCD_TOTAL_VISIBLE_PAGES];
+uint8_t STLcdSyncBuffer[LCD_PAGE_LEN * LCD_TOTAL_VISIBLE_PAGES];
+
+
+// Bounding box.
+static uint8_t STLcdUpdateXMin = LCD_WIDTH, STLcdUpdateXMax = 0, STLcdUpdateYMin = LCD_HEIGHT, STLcdUpdateYMax = 0;
+static uint8_t STLcdSync = 0;
+static uint8_t STLcdSyncPageMin, STLcdSyncPageMax, STLcdSyncColumnMin, STLcdSyncColumnMax;
+
+static void STLcd_updateBoundingBox(uint8_t xmin, uint8_t ymin, uint8_t xmax, uint8_t ymax) {
+    if (xmin < STLcdUpdateXMin) STLcdUpdateXMin = xmin;
+    if (xmax > STLcdUpdateXMax) STLcdUpdateXMax = xmax;
+    if (ymin < STLcdUpdateYMin) STLcdUpdateYMin = ymin;
+    if (ymax > STLcdUpdateYMax) STLcdUpdateYMax = ymax;
+}
+
+void STLcd_clear(void) {
+    memset(STLcdBuffer, 0, LCD_PAGE_LEN * LCD_TOTAL_VISIBLE_PAGES);
+    STLcd_updateBoundingBox(0, 0, LCD_WIDTH, LCD_HEIGHT);
+}
+
+static uint8_t STLcdSyncPageCurrent;
+static uint8_t *STLcdSyncBufferColumnCurrent;
+static uint8_t *STLcdSyncBufferColumnMin, *STLcdSyncBufferColumnMax;
+static uint32_t STLcdSyncStage;
+
+#define STAGE_WRITE_START 15
+#define STAGE_WRITE_END 20
+
+void STLcd_blockingSync(void) {
+    for(;STLcdSyncPageCurrent < STLcdSyncPageMax;
+	STLcdSyncPageCurrent++, STLcdSyncBufferColumnMin += LCD_PAGE_LEN)
+    {
+	LCD_writeControlReg(0xB0 | STLcdSyncPageCurrent);
+	LCD_writeControlReg(0x10 | STLcdSyncColumnMin >> 4);
+	LCD_writeControlReg(0x10 | (STLcdSyncColumnMin & 0x0f));
+	SPI_write(STLcdSyncBufferColumnMin, STLcdSyncColumnMax - STLcdSyncColumnMin);
+    }
+}
+
+// a non-blocking sync function
+void STLcd_sync(void) {
```

Comment:
(As usual, tabs).

This looks really awesome! I need to do something similar to this for other things on the keyboard.
One thing to make it easier to read (and possibly fix future bugs).

Put a one liner (or more), above each case statement about what *you* think each piece does. If you're not quite sure, just put a TODO instead and I can fill in the comment later.

For example:
`// Verify SPI0 TxFIFO is not full, then enable LCD configuration registers (A0 to Low)
case 0:`